File: game/src/player.rs
//! Game project.
use fyrox::asset::Resource;
use fyrox::core::algebra::{Rotation3, UnitQuaternion};
use fyrox::core::num_traits::Zero;
use fyrox::core::ComponentProvider;
use fyrox::event::{DeviceEvent, MouseButton};
use fyrox::resource::model::Model;
use fyrox::{
    core::{
        algebra::Vector3, pool::Handle, reflect::prelude::*, type_traits::prelude::*,
        visitor::prelude::*, TypeUuidProvider,
    },
    event::{ElementState, Event, WindowEvent},
    keyboard::{KeyCode, PhysicalKey},
    scene::node::Node,
    script::{ScriptContext, ScriptTrait},
};
use fyrox_lite_api::lite_ctx::{LiteContext, LiteScript};
use fyrox_lite_api::lite_node::LiteNode;
use fyrox_lite_api::lite_window::LiteWindow;
use std::f32::consts::PI;
use std::ops::Mul;

use crate::bullet::{Bullet, BulletHit, BulletSeed};
use crate::fyrox_utils::HandleNodeExt;
use crate::game::Game;

#[derive(Visit, Reflect, Debug, Clone, TypeUuidProvider, ComponentProvider, Default)]
#[type_uuid(id = "c5671d19-9f1a-4286-8486-add4ebaadaec")]
#[visit(optional)]
pub struct Player {
    sensitivity: f32,
    camera: Handle<Node>,
    power: f32,
    bullet: Option<Resource<Model>>,
    initial_bullet_velocity: f32,
    shooting_range: f32,
    reload_delay_sec: f32,

    #[reflect(hidden)]
    reload_sec: f32,

    #[reflect(hidden)]
    published: bool,

    #[reflect(hidden)]
    collider: Handle<Node>,

    #[visit(skip)]
    #[reflect(hidden)]
    temp: TempState,
}

#[derive(Debug, Default, Clone)]
pub struct TempState {
    pub aim_y: f32,
    pub forward: bool,
    pub back: bool,
    pub left: bool,
    pub right: bool,
    pub fire: bool,
}

impl Player {
    fn turn(&self, x: f32, ctx: &mut LiteContext) {
        let rot_delta = Rotation3::from_axis_angle(&Vector3::y_axis(), self.sensitivity * x);
        ctx.node
            .set_local_rotation(ctx.node.local_rotation().mul(&rot_delta));
    }

    fn aim(&mut self, y: f32) {
        self.temp.aim_y += y * self.sensitivity;

        self.temp.aim_y = self.temp.aim_y.clamp(-PI / 2.0, PI / 2.0);

        LiteNode::from(self.camera).set_local_rotation(UnitQuaternion::from_axis_angle(
            &Vector3::x_axis(),
            self.temp.aim_y,
        ));
    }

    fn fire(&mut self) {
        let camera_pos = LiteNode::from(self.camera).global_position();
        let bullet_orientation = LiteNode::from(self.camera).global_rotation();

        let prefab = self.bullet.as_ref().unwrap().clone();
        Bullet::spawn(BulletSeed {
            prefab,
            origin: camera_pos,
            direction: bullet_orientation.transform_vector(&Vector3::z_axis()),
            initial_velocity: self.initial_bullet_velocity,
            author_collider: self.collider,
            range: self.shooting_range,
        });
        println!("bullet spawned");
    }
}

impl LiteScript for Player {
    fn on_init(&mut self, ctx: &mut LiteContext) {
        let _ = LiteWindow::set_cursor_grab(fyrox::window::CursorGrabMode::Confined);

        self.collider = ctx
            .node
            .try_get_collider()
            .expect("Collider not found under Player node")
            .into();
    }

    fn on_start(&mut self, ctx: &mut LiteContext) {
        ctx.node.subscribe_to::<BulletHit>();
    }

    fn on_message(
        &mut self,
        message: &mut dyn fyrox::script::ScriptMessagePayload,
        ctx: &mut fyrox_lite_api::lite_node::LiteMessageContext,
    ) {
        if let Some(_bullet) = message.downcast_ref::<BulletHit>() {
            ctx.with_plugin::<Game, _>(|it| it.wounds += 1);
            println!("player wounded!");
        }
    }

    fn on_update(&mut self, ctx: &mut LiteContext) {
        profiling::scope!("Player::on_update");
        if self.reload_sec > 0.0 {
            self.reload_sec -= ctx.dt;
        }
        if !self.published {
            self.published = true;
            let player = ctx.node;
            ctx.with_plugin::<Game, _>(|it| it.player = player.into());
        }

        if self.temp.fire {
            if self.reload_sec <= 0.0 {
                self.reload_sec = self.reload_delay_sec;
                self.fire();
            }
        }

        let mut move_delta = Vector3::<f32>::zero();
        if self.temp.forward {
            move_delta.z += 1.0
        }
        if self.temp.back {
            move_delta.z -= 1.0
        }
        if self.temp.left {
            move_delta.x += 1.0
        }
        if self.temp.right {
            move_delta.x -= 1.0
        }

        if move_delta.magnitude() > 0.001 {
            move_delta.normalize_mut();
        }

        let self_rotation = ctx.node.local_rotation().clone();
        let move_delta = self_rotation.transform_vector(&move_delta);
        let force = move_delta * self.power;
        ctx.node.with_rigid_body(|it| it.apply_force(force));
    }

    fn on_os_event(&mut self, event: &Event<()>, ctx: &mut LiteContext) {
        if let Event::WindowEvent { event, .. } = event {
            if let WindowEvent::KeyboardInput {
                device_id: _,
                event,
                is_synthetic: _,
            } = event
            {
                let value = match event.state {
                    ElementState::Pressed => true,
                    ElementState::Released => false,
                };
                match event.physical_key {
                    PhysicalKey::Code(KeyCode::KeyW) => self.temp.forward = value,
                    PhysicalKey::Code(KeyCode::KeyS) => self.temp.back = value,
                    PhysicalKey::Code(KeyCode::KeyA) => self.temp.left = value,
                    PhysicalKey::Code(KeyCode::KeyD) => self.temp.right = value,
                    _ => (),
                }
            }
            if let WindowEvent::MouseInput {
                device_id: _,
                state,
                button,
            } = event
            {
                if *button == MouseButton::Left {
                    self.temp.fire = match state {
                        ElementState::Pressed => true,
                        ElementState::Released => false,
                    };
                }
            }
        }
        if let Event::DeviceEvent {
            device_id: _,
            event,
        } = event
        {
            if let DeviceEvent::MouseMotion { delta: (x, y) } = event {
                self.turn(-*x as f32, ctx);
                self.aim(*y as f32);
            }
        }
    }
}

impl ScriptTrait for Player {
    fn on_init(&mut self, #[allow(unused_variables)] ctx: &mut ScriptContext) {
        self.redispatch_init(ctx);
    }

    fn on_start(&mut self, #[allow(unused_variables)] ctx: &mut ScriptContext) {
        self.redispatch_start(ctx);
    }

    fn on_message(
        &mut self,
        message: &mut dyn fyrox::script::ScriptMessagePayload,
        ctx: &mut fyrox::script::ScriptMessageContext,
    ) {
        self.redispatch_message(message, ctx);
    }

    fn on_update(&mut self, ctx: &mut ScriptContext) {
        self.redispatch_update(ctx);
    }

    fn on_os_event(&mut self, event: &Event<()>, ctx: &mut ScriptContext) {
        self.redispatch_os_event(event, ctx);
    }
}

File: scripts/Player.lua
---@uuid c5671d19-9f1a-4286-8486-add4ebaadaec
---@class Player : Script
---@field sensitivity number
---@field camera Node
---@field power number
---@field bullet Prefab
---@field initial_bullet_velocity number
---@field shooting_range number
---@field reload_delay_sec number
---@field private reload_sec number
---@field private published boolean
---@field private collider Node
Player = script_class()

local aim_y = 0
local forward = false
local back = false
local left = false
local right = false
local fire = false


---@param x number
function Player:turn(x)
    local rot_delta = Quaternion:from_axis_angle(Vector3.Y, self.sensitivity * x);

    self.node:set_local_rotation(self.node:local_rotation():mul(rot_delta))
end

function Player:aim(y)
    aim_y = aim_y + y * self.sensitivity;

    aim_y = math.max(-math.pi / 2.0, math.min(aim_y, math.pi / 2.0))

    self.camera:set_local_rotation(Quaternion:from_axis_angle(
        Vector3.X,
        aim_y
    ));
end

function Player:fire()
    local camera_pos = self.camera:global_position();

    local bullet_orientation = self.camera:global_rotation();

    Bullet:spawn(
        {
            prefab = self.bullet,
            origin = camera_pos,
            direction = bullet_orientation:mul(Vector3.Z),
            initial_velocity = self.initial_bullet_velocity,
            author_collider = self.collider,
            range = self.shooting_range
        }
    );
end


function Player:on_init()
    Window:set_cursor_grab(CursorGrabMode.Confined);

    self.collider = self.node:find_collider_in_children()
    end

function Player:on_start(ctx)
    self.node:subscribe_to()
end

function Player:on_message(message, ctx)
    local _bullet = message.downcast_ref("BulletHit")
    if _bullet ~= nil then
        Plugin:get("Game"):inc_wounds()
        print("player wounded!")
    end
end

function Player:on_update(dt)
    if self.reload_sec > 0.0 then
        self.reload_sec = self.reload_sec - dt;
    end
    if not self.published then
        self.published = true;
        Plugin:get("Game").player = self.node;
    end

    if fire then
        if self.reload_sec <= 0.0 then
            self.reload_sec = self.reload_delay_sec;
            self:fire();
        end
    end

    local move_delta = Vector3.ZERO;
    if forward then
        move_delta.z = move_delta.z + 1.0
    end
    if back then
        move_delta.z = move_delta.z - 1.0
    end
    if left then
        move_delta.x = move_delta.x + 1.0
    end
    if right then
        move_delta.x = move_delta.x - 1.0
    end

    if move_delta:magnitude() > 0.001 then
        move_delta:normalize_inplace();
    end

    local self_rotation = self.node:local_rotation()
    local move_delta = self_rotation:mul(move_delta);
    local force = move_delta:mul(self.power);
    self.node:as_rigid_body():apply_force(force);
end

function Player:on_os_event(event)
    if event:is("WindowEvent") then
        local event = event.event
        if event:is("KeyboardInput") then
            local event = event.event
            local value = event.state:is("Pressed")
            if event.physical_key:is("Code") then
                local code = event.physical_key[1]
                if code:is("KeyW") then
                    forward = value
                end
                if code:is("KeyS") then
                    back = value
                end
                if code:is("KeyA") then
                    left = value
                end
                if code:is("KeyD") then
                    right = value
                end
            end
        end
        if event:is("MouseInput") then
            if event.button:is("Left") then
                fire = event.state:is("Pressed")
            end
        end
    end
    if event:is("DeviceEvent") then
        local event = event.event
        if event:is("MouseMotion") then
            local x, y = table.unpack(event.delta)
            self:turn(-x);
            self:aim(y);
        end
    end
end
